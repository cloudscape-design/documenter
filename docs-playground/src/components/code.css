.code {
  background: transparent;
  margin: 0;
  padding: 0;
  font-family: monospace;
  font-size: 1em;
}

.code-token-link {
  font-family: monospace;
  font-size: 1em;
}

.code-token-symbol {
  /* $color-text-body-secondary */
  color: #545b64;
}
