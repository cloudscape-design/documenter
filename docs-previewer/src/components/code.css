/* 
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
  SPDX-License-Identifier: Apache-2.0
*/

.code {
  background: transparent;
  margin: 0;
  padding: 0;
  font-family: monospace;
  font-size: 1em;
}

.code-token-link {
  font-family: monospace;
  font-size: 1em;
}

.code-token-symbol {
  /* $color-text-body-secondary */
  color: #545b64;
}
