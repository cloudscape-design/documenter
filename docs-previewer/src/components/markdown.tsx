// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import Link from '@cloudscape-design/components/link';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { Remarkable } from 'remarkable';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import RemarkableReactRenderer from 'remarkable-react';

const md = new Remarkable();
md.renderer = new RemarkableReactRenderer({
  components: {
    a: Link,
  },
});

export default function Markdown({ children }: { children: string }): JSX.Element {
  return md.render(children);
}
