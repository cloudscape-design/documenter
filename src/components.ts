// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import * as path from 'node:path';
import * as assert from 'node:assert/strict';
import ts from 'typescript';
import { glob } from 'fast-glob';
import { pascalCase } from 'change-case';

import {
  ComponentDefinition,
  ComponentFunction,
  ComponentProperty,
  ComponentRegion,
  EventHandler,
  InlineTypeDefinition,
} from './interfaces';

import {
  nodeToString,
  typeToString,
  joinSymbolDisplayParts,
  isTruthy,
  sortSymbolsByName,
  signatureDeclarationToString,
} from './util';

const EVENT_HANDLER_REGEX = /^on[A-Z]/;
const PASSTHROUGH_TYPE_NAMES = ['HTMLElement'];

export function documentComponents(tsconfigPath: string, publicFilesGlob: string): ComponentDefinition[] {
  // HACK: assuming "src" directory cause it's easier for now
  const allFiles = glob.sync(path.join(path.dirname(tsconfigPath), 'src/**/*.{ts,tsx}'), { absolute: true });
  const componentEntryFiles = glob.sync(publicFilesGlob, { absolute: true });

  // Don't provide the tsconfig because it unpredictably alters how TypeScript parses source files.
  // Thankfully, we don't really need much config-specific stuff here.
  const program = ts.createProgram(allFiles, { strict: true });

  return program
    .getSourceFiles()
    .filter(({ fileName }) => componentEntryFiles.includes(fileName))
    .flatMap(sourceFile => buildComponentDefinition(program, sourceFile))
    .filter(isTruthy);
}

export function buildComponentDefinition(
  program: ts.Program,
  sourceFile: ts.SourceFile,
): ComponentDefinition | undefined {
  /** The expected component name, based on the naming convention of the folder. */
  const componentName = pascalCase(path.basename(path.dirname(sourceFile.fileName)));

  /** The expected props type name, based on the naming convention of the folder. */
  const propsName = `${componentName}Props`;

  /** All the types and values exported from this source file. */
  const fileExports = program.getTypeChecker().getSymbolAtLocation(sourceFile)?.exports;
  if (!fileExports) {
    console.warn(`File has no exports! ${sourceFile.fileName}`);
    return;
  }

  /**
   * The AST node where the component was most likely defined.
   * This node's parameters can inspected to fetch default property values.
   */
  let componentNode: ts.SignatureDeclaration | undefined;

  /** The likely interface/type of the component properties. */
  let propsType: ts.Type | undefined;

  fileExports.forEach((tsSymbol, exportedName) => {
    if (exportedName === 'default') {
      // Get the non-type declaration that matches this export.
      // It's not possible to have two default exports, so only one is a non-type node.
      const node = tsSymbol.getDeclarations()?.find(node => !ts.isTypeNode(node));
      componentNode = node && getComponentFunctionNode(program, node);
    }

    if (exportedName === propsName) {
      propsType = program.getTypeChecker().getDeclaredTypeOfSymbol(tsSymbol);
    }
  });

  assert.ok(componentNode, `File has no default non-type export: ${sourceFile.fileName}`);
  assert.ok(propsType, `File has no props export (${propsName}): ${sourceFile.fileName}`);

  //
  // Get component documentation tags
  //

  let componentVersion: string | undefined;
  let componentDoc: string | undefined;
  let componentBeta = false;

  const componentSymbol = program.getTypeChecker().getSymbolAtLocation(componentNode.name ?? componentNode);
  if (componentSymbol) {
    const versionTag = componentSymbol.getJsDocTags().find(tag => tag.name === 'version');
    componentVersion = versionTag?.text && joinSymbolDisplayParts(versionTag.text);
    componentDoc =
      joinSymbolDisplayParts(componentSymbol.getDocumentationComment(program.getTypeChecker())) || undefined;
    componentBeta = componentSymbol.getJsDocTags().some(tag => tag.name === 'beta');
  }

  //
  // Extract default values from componentNode
  //

  const defaultValues = new Map<string, string>();

  if (componentNode.parameters[0] && ts.isObjectBindingPattern(componentNode.parameters[0].name)) {
    for (const bindingElement of componentNode.parameters[0].name.elements) {
      if (ts.isIdentifier(bindingElement.name) && bindingElement.initializer) {
        const defaultValueString = nodeToString(sourceFile, bindingElement.initializer, ts.EmitHint.Expression);
        defaultValues.set(bindingElement.name.getText(), defaultValueString);
      }
    }
  }

  //
  // Find a type under the props type named "Ref" and pull the functions from there.
  //
  const functions: Array<ComponentFunction> = [];

  const refSymbol = propsType.getSymbol()?.exports?.get(ts.escapeLeadingUnderscores('Ref'));
  const refProps = refSymbol && program.getTypeChecker().getDeclaredTypeOfSymbol(refSymbol).getProperties();
  for (const fnSymbol of refProps ?? []) {
    const refFunctionType = program.getTypeChecker().getTypeOfSymbol(fnSymbol);
    // stripUnionWithUndefined because optional functions are treated as being unioned with undefined.
    // FIXME: Should ref functions even be optional? We don't document them as optional except in typings.
    const funcSignature = stripUnionWithUndefined(refFunctionType).getCallSignatures()[0];
    assert.ok(funcSignature, `Could not find function signature for ${fnSymbol.getName()}`);

    functions.push({
      name: fnSymbol.getName(),
      description: joinSymbolDisplayParts(fnSymbol.getDocumentationComment(program.getTypeChecker())),
      parameters: funcSignature.parameters.map(parameter => ({
        name: parameter.getName(),
        type: typeToString(program, program.getTypeChecker().getTypeOfSymbol(parameter)),
      })),
      returnType: typeToString(program, program.getTypeChecker().getReturnTypeOfSignature(funcSignature)),
    });
  }

  //
  // Extract properties from propsType
  //

  const properties: Array<ComponentProperty> = [];
  const regions: Array<ComponentRegion> = [];
  const events: Array<EventHandler> = [];

  const interfaceProps = propsType.getProperties();
  assert.ok(interfaceProps, `No properties found for ${propsName}`);

  for (const prop of sortSymbolsByName(interfaceProps)) {
    const declaration = prop.getDeclarations()?.[0];
    assert.ok(declaration && ts.isPropertySignature(declaration) && declaration.type);

    if (declaration.type.getText() === 'React.ReactNode' || declaration.type.getText() === 'ReactNode') {
      // If a property accepts React.ReactNode, it's treated as a slot.
      const visualRefreshTag = prop.getJsDocTags().find(tag => tag.name === 'visualrefresh');
      regions.push({
        name: prop.getName(),
        description: joinSymbolDisplayParts(prop.getDocumentationComment(program.getTypeChecker())) || undefined,
        displayName: joinSymbolDisplayParts(prop.getJsDocTags().find(tag => tag.name === 'displayname')?.text),
        deprecatedTag: joinSymbolDisplayParts(prop.getJsDocTags().find(tag => tag.name === 'deprecated')?.text),
        visualRefreshTag: visualRefreshTag ? joinSymbolDisplayParts(visualRefreshTag.text) ?? '' : undefined,
        i18nTag: prop.getJsDocTags().some(tag => tag.name === 'i18n') || undefined,
        inlineType: undefined,
        isDefault: prop.getName() === 'children',
      });
    } else if (EVENT_HANDLER_REGEX.test(prop.getName())) {
      // Event handler types are expected to be either CancelableEventHandler<T> or NonCancelableEventHandler<T>.
      // The <T> is what's documented in detailType and detailInlineType.
      let detailType: ts.Type | undefined;
      let detailTypeString: string | undefined;
      let isCancelable = false;

      if (ts.isTypeReferenceNode(declaration.type)) {
        const detailTypeNode = declaration.type.typeArguments?.[0];
        if (detailTypeNode) {
          detailType = program.getTypeChecker().getTypeFromTypeNode(detailTypeNode);
          detailTypeString = nodeToString(sourceFile, detailTypeNode, ts.EmitHint.Unspecified);
        }
        isCancelable = declaration.type.typeName.getText() !== 'NonCancelableEventHandler';
      }

      events.push({
        name: prop.getName(),
        description: joinSymbolDisplayParts(prop.getDocumentationComment(program.getTypeChecker())) || undefined,
        deprecatedTag: joinSymbolDisplayParts(prop.getJsDocTags().find(tag => tag.name === 'deprecated')?.text),
        cancelable: isCancelable,
        detailType: detailTypeString,
        detailInlineType:
          detailType && detailTypeString ? buildInlineTypeDefinition(program, detailTypeString, detailType) : undefined,
      });
    } else {
      const propType = program.getTypeChecker().getTypeFromTypeNode(declaration.type);
      const typeString = typeToString(program, propType);
      const visualRefreshTag = prop.getJsDocTags().find(tag => tag.name === 'visualrefresh');

      const property: ComponentProperty = {
        name: prop.getName(),
        type: typeString,
        description: joinSymbolDisplayParts(prop.getDocumentationComment(program.getTypeChecker())) || undefined,
        deprecatedTag: joinSymbolDisplayParts(prop.getJsDocTags().find(tag => tag.name === 'deprecated')?.text),
        visualRefreshTag: visualRefreshTag ? joinSymbolDisplayParts(visualRefreshTag.text) ?? '' : undefined,
        i18nTag: prop.getJsDocTags().some(tag => tag.name === 'i18n') || undefined,
        optional: !!(prop.getFlags() & ts.SymbolFlags.Optional) || isUnionWithUndefined(propType),
        defaultValue: defaultValues.get(prop.getName()),
      };

      if (ts.isTypeReferenceNode(declaration.type)) {
        // This is a named type (ThingProps.Item). Generally, we expect all
        // property types to be defined this way, for developer convenience.
        property.inlineType = buildInlineTypeDefinition(program, typeString, propType);

        // For compatibility, a union of strings should always be just "string".
        if (isUnionOfStrings(propType)) {
          property.type = 'string';
        }
      } else if (isInlineUnionOfStrings(declaration.type)) {
        // This is a union type, but is defined inline.
        // The previous documenter documented this as an inline type with an empty name.
        property.type = 'string';
        property.inlineType = {
          name: '',
          type: 'union',
          values: declaration.type.types.map(t => ((t as ts.LiteralTypeNode).literal as ts.StringLiteral).text).sort(),
        };
      }

      properties.push(property);
    }
  }

  return {
    name: componentName,
    version: componentVersion,
    description: componentDoc,
    releaseStatus: componentBeta ? 'beta' : 'stable',
    properties,
    regions,
    events,
    functions,
  };
}

function getComponentFunctionNode(program: ts.Program, startNode: ts.Node): ts.SignatureDeclaration {
  let node: ts.Node | undefined = startNode;

  if (ts.isExportAssignment(node)) {
    // This is an "export default Abc" statement. This gets the node where "Abc" is declared.
    node = program.getTypeChecker().getSymbolAtLocation(node.expression)?.valueDeclaration;
    assert.ok(node);
  }

  if (ts.isVariableDeclaration(node)) {
    // This is a "const Abc = ..." statement.
    // For simplicity, we require that a variable be declared and initialized together.
    // NOTE: If needed, we'll need to pull documentation from here, not the function node.
    assert.ok(node.initializer);
    node = node.initializer;
  }

  if (ts.isAsExpression(node)) {
    // This is a "value as Type" statement.
    // NOTE: For now, we discard the type information and use the naming convention. Maybe we shouldn't?
    node = node.expression;
  }

  if (ts.isCallExpression(node)) {
    // Function call, very likely React.memo or React.forwardRef.
    // NOTE: Actually validate that this is a known/expected function?
    assert.ok(node.arguments[0]);
    node = node.arguments[0];
  }

  assert.ok(ts.isFunctionLike(node), 'Unexpected default export ts.SyntaxKind: ' + node.kind);
  return node;
}

function buildInlineTypeDefinition(program: ts.Program, name: string, type: ts.Type): InlineTypeDefinition | undefined {
  // If it's a "global" type (e.g. HTMLElement), we don't want to turn it
  // into an inline type. So we keep a blocklist for now. There may be a
  // better way to do this.
  if (PASSTHROUGH_TYPE_NAMES.includes(name)) {
    return undefined;
  }

  if (type.isUnion()) {
    return {
      name,
      type: 'union',
      values: type.types.map(t => typeToString(program, t, true)).sort(),
    };
  }

  // If a type has a call signature (even if it's defined as an object),
  // we treat it as a "function". We have no use for callable types that
  // also have properties.
  const callSignatures = type.getCallSignatures();
  if (callSignatures.length > 0) {
    const signature = callSignatures[0];
    return {
      name,
      type: 'function',
      parameters: signature.parameters.map(parameter => {
        const declaration = parameter.getDeclarations()?.[0];
        assert.ok(declaration && ts.isParameter(declaration) && declaration.type);
        return {
          name: parameter.getName(),
          type: typeToString(program, program.getTypeChecker().getTypeFromTypeNode(declaration.type)),
        };
      }),
      returnType: typeToString(program, program.getTypeChecker().getReturnTypeOfSignature(signature)),
    };
  }

  // We can treat non-union, non-callable types as object types (or types that
  // resolve down to object types). One exception is arrays, which are technically
  // object types, but we don't want JS array methods (like sort) to end up in here.
  const hasIndexSignatures = !!type.getNumberIndexType() || !!type.getStringIndexType();
  if (!type.isLiteral() && !hasIndexSignatures && type.getProperties().length > 0) {
    return {
      name,
      type: 'object',
      properties: sortSymbolsByName(type.getProperties()).map(property => {
        const declaration = property.getDeclarations()?.[0];
        assert.ok(declaration);

        if (ts.isPropertySignature(declaration)) {
          assert.ok(declaration.type);
          const type = program.getTypeChecker().getTypeFromTypeNode(declaration.type);
          return {
            name: property.getName(),
            optional: !!(property.flags & ts.SymbolFlags.Optional) || isUnionWithUndefined(type),
            type: typeToString(program, type),
          };
        }

        if (ts.isMethodSignature(declaration)) {
          return {
            name: property.getName(),
            optional: !!(property.flags & ts.SymbolFlags.Optional) || isUnionWithUndefined(type),
            type: signatureDeclarationToString(program, declaration),
          };
        }

        assert.fail(`Unexpected declaration kind: ${declaration.kind}`);
      }),
    };
  }
}

function isInlineUnionOfStrings(type: ts.TypeNode): type is ts.UnionTypeNode {
  return (
    ts.isUnionTypeNode(type) &&
    type.types.every(subType => ts.isLiteralTypeNode(subType) && ts.isStringLiteral(subType.literal))
  );
}

function isUnionOfStrings(type: ts.Type): type is ts.UnionType {
  return type.isUnion() && type.types.every(subType => subType.isStringLiteral());
}

function isUnionWithUndefined(type: ts.Type): type is ts.UnionType {
  return type.isUnion() && type.types.length === 2 && type.types.some(t => t.getFlags() & ts.TypeFlags.Undefined);
}

function stripUnionWithUndefined(type: ts.Type): ts.Type {
  if (isUnionWithUndefined(type)) {
    return type.types.find(subtype => (subtype.getFlags() & ts.TypeFlags.Undefined) === 0)!;
  }
  return type;
}
