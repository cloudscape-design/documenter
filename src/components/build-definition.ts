// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { DeclarationReflection, ReflectionKind } from 'typedoc';
import { Type } from 'typedoc/dist/lib/models';
import { ComponentDefinition, ComponentFunction } from './interfaces';
import schema from '../schema';
import buildTypeDefinition from './build-type-definition';
import extractDefaultValues from './default-values-extractor';

function buildEventInfo(handler: DeclarationReflection) {
  if (!schema.types.isReferenceType(handler.type)) {
    throw new Error(
      `Unknown event handler type: ${handler.type && handler.type.type} at ${schema.utils.getDeclarationSourceFilename(
        handler
      )}`
    );
  }
  const detailType = handler.type.typeArguments?.[0];
  const { typeName, typeDefinition } = detailType
    ? getPropertyType(detailType)
    : { typeName: undefined, typeDefinition: undefined };
  return {
    name: handler.name,
    description: schema.code.buildNodeDescription(handler),
    cancelable: handler.type.name !== 'NonCancelableEventHandler',
    detailType: typeName,
    detailInlineType: typeDefinition,
    deprecatedTag: handler.comment?.tags?.find(tag => tag.tagName === 'deprecated')?.text.trim(),
  };
}

function buildMethodsDefinition(refType?: DeclarationReflection): ComponentFunction[] {
  if (!refType || !refType.children) {
    return [];
  }
  return refType.children.map(child => {
    if (!child.signatures) {
      throw new Error(
        `${schema.code.buildFullName(child)} should contain only methods, "${
          child.name
        }" has a "${schema.code.buildType(child.type)}" type`
      );
    }
    if (child.signatures.length > 1) {
      throw new Error(
        `Method overloads are not supported, found multiple signatures at ${schema.utils.getDeclarationSourceFilename(
          child
        )}`
      );
    }
    const signature = child.signatures[0];
    return {
      name: child.name,
      description: schema.code.buildNodeDescription(signature),
      returnType: schema.code.buildType(signature.type),
      parameters:
        signature.parameters?.map(parameter => ({
          name: parameter.name,
          type: schema.code.buildType(parameter.type),
        })) ?? [],
    };
  });
}

function getPropertyType(type?: Type) {
  const typeAlias = schema.types.isReferenceType(type) && (type.reflection as DeclarationReflection | undefined);
  const resolvedType = typeAlias ? typeAlias.type : type;
  if (schema.types.isUnionType(resolvedType)) {
    const subTypes = schema.utils.excludeUndefinedTypeFromUnion(resolvedType);
    if (subTypes.length > 1) {
      if (subTypes.every(type => schema.types.isStringLiteralType(type))) {
        const referenceTypeName = schema.types.isReferenceType(type) ? schema.code.buildType(type) : '';
        const declaration = new DeclarationReflection(referenceTypeName, ReflectionKind.TypeLiteral);
        declaration.type = resolvedType;
        return {
          typeName: 'string',
          typeDefinition: buildTypeDefinition(declaration),
        };
      }
      if (
        subTypes.every(type => schema.types.isIntrinsicType(type) && (type.name === 'true' || type.name === 'false'))
      ) {
        return { typeName: 'boolean' };
      }
    }
  }
  return {
    typeName: schema.code.buildType(type),
    typeDefinition: typeAlias ? buildTypeDefinition(typeAlias) : undefined,
  };
}

export default function buildDefinition(
  component: DeclarationReflection,
  props: DeclarationReflection[],
  objects: DeclarationReflection[]
): ComponentDefinition {
  const events = props.filter(prop => prop.name.match(/^on[A-Z]/));
  const regions = props.filter(prop => ['React.ReactNode', 'ReactNode'].includes(schema.code.buildType(prop.type)));
  const onlyProps = props.filter(prop => !events.includes(prop) && !regions.includes(prop));
  const defaultValues = extractDefaultValues(component);

  const betaTag = component.signatures && component.signatures[0].comment?.tags?.find(tag => tag.tagName === 'beta');
  const versionTag =
    component.signatures && component.signatures[0].comment?.tags?.find(tag => tag.tagName === 'version');

  return {
    name: component.name,
    version: versionTag?.text.trim().replace('\n', ''),
    releaseStatus: betaTag ? 'beta' : 'stable',
    description: schema.code.buildDeclarationDescription(component),
    regions: regions.map(region => {
      return {
        name: region.name,
        displayName: region.comment?.tags?.find(tag => tag.tagName === 'displayname')?.text.trim(),
        description: schema.code.buildNodeDescription(region),
        isDefault: region.name === 'children',
        visualRefreshTag: region.comment?.tags?.find(tag => tag.tagName === 'visualrefresh')?.text.trim(),
        deprecatedTag: region.comment?.tags?.find(tag => tag.tagName === 'deprecated')?.text.trim(),
        i18nTag: region.comment?.tags?.find(tag => tag.tagName === 'i18n')?.text.trim(),
      };
    }),
    functions: buildMethodsDefinition(objects.find(def => def.name === 'Ref')),
    properties: onlyProps.map(prop => {
      const { typeName, typeDefinition } = getPropertyType(prop.type);
      return {
        name: prop.name,
        type: typeName,
        inlineType: typeDefinition,
        optional: schema.utils.isOptionalDeclaration(prop),
        description: schema.code.buildNodeDescription(prop),
        defaultValue: defaultValues[prop.name],
        visualRefreshTag: prop.comment?.tags?.find(tag => tag.tagName === 'visualrefresh')?.text.trim(),
        deprecatedTag: prop.comment?.tags?.find(tag => tag.tagName === 'deprecated')?.text.trim(),
        i18nTag: prop.comment?.tags?.find(tag => tag.tagName === 'i18n')?.text.trim(),
      };
    }),
    events: events.map(handler => buildEventInfo(handler)),
  };
}
