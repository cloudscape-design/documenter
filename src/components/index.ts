// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { pascalCase } from 'change-case';
import pathe from 'pathe';
import { matcher } from 'micromatch';

import { buildComponentDefinition } from './component-definition';
import { extractDefaultValues, extractExports, extractFunctions, extractProps } from './extractor';
import type { ComponentDefinition } from './interfaces';
import { bootstrapTypescriptProject } from '../bootstrap/typescript';
import { extractDeclaration, getDescription } from './type-utils';

function componentNameFromPath(componentPath: string) {
  const dashCaseName = pathe.basename(pathe.dirname(componentPath));
  return { dashCaseName, name: pascalCase(dashCaseName) };
}

export interface DocumenterOptions {
  tsconfigPath: string;
  publicFilesGlob: string;
  extraExports?: Record<string, Array<string>>;
}

export function documentComponents(options: DocumenterOptions): Array<ComponentDefinition> {
  const program = bootstrapTypescriptProject(options.tsconfigPath);
  const checker = program.getTypeChecker();

  const isMatch = matcher(pathe.resolve(options.publicFilesGlob));

  return program
    .getSourceFiles()
    .filter(file => isMatch(file.fileName))
    .map(sourceFile => {
      const moduleSymbol = checker.getSymbolAtLocation(sourceFile);
      const { name, dashCaseName } = componentNameFromPath(sourceFile.fileName);

      // istanbul ignore next
      if (!moduleSymbol) {
        throw new Error(`Unable to resolve module: ${sourceFile.fileName}`);
      }
      const exportSymbols = checker.getExportsOfModule(moduleSymbol);
      const { propsSymbol, componentSymbol } = extractExports(
        name,
        exportSymbols,
        checker,
        options?.extraExports ?? {}
      );
      const props = extractProps(propsSymbol, checker);
      const functions = extractFunctions(propsSymbol, checker);
      const defaultValues = extractDefaultValues(componentSymbol, checker);
      const componentDescription = getDescription(
        componentSymbol.getDocumentationComment(checker),
        extractDeclaration(componentSymbol)
      );

      return buildComponentDefinition(
        name,
        dashCaseName,
        props,
        functions,
        defaultValues,
        componentDescription,
        checker
      );
    });
}
