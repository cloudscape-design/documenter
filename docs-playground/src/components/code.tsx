// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import React from 'react';
import Link from '@cloudscape-design/components/link';

import './code.css';

export function Code({ children }: { children: React.ReactNode }) {
  return <pre className="code">{children}</pre>;
}

export function CodeToken({
  children,
  kind,
  options,
  onNavigate,
}: {
  children: React.ReactNode;
  kind: string;
  options: any;
  onNavigate: (path: string[]) => void;
}) {
  switch (kind) {
    case 'whitespace':
      return (
        <>
          <Token kind="whitespace">{children}</Token>
          <wbr />
        </>
      );
    case 'name':
      return options?.path ? (
        <Link onFollow={() => onNavigate(options?.path)} className="code-token-link">
          <Token kind="name">{children}</Token>
        </Link>
      ) : (
        <Token kind={kind}>{children}</Token>
      );
    default:
      return <Token kind={kind}>{children}</Token>;
  }
}

function Token({ children, kind }: { children: React.ReactNode; kind: string }) {
  return <span className={`code-token-${kind}`}>{children}</span>;
}
