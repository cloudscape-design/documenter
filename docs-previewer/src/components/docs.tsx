// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import React, { useEffect, useRef, useState } from 'react';
import { SideNavigation, Box, SpaceBetween, Header, Link, Badge, Table } from '@cloudscape-design/components';
import { Code, CodeToken } from './code';
import Markdown from './markdown';
import PageLayout from '../page-layout';
import { useNavigate, Routes, Route, useLocation } from 'react-router-dom';

type Definition = any;

interface NodeProps {
  definition: Definition;
  onNavigate: (name: string) => void;
  level: number;
}

export default function Docs({
  definitions,
  namespace,
}: {
  definitions: { name: string; folder: string; promise: Promise<Definition> }[];
  namespace: string;
}): JSX.Element {
  const navigate = useNavigate();

  const onNavigate = (path: string) => navigate(`/${namespace}/${path}`);
  const uniqueNames = Array.from(new Set(definitions.map(it => it.name)));

  return (
    <PageLayout
      navigation={
        <SideNavigation
          header={{ href: '/', text: 'Index' }}
          onFollow={e => {
            e.preventDefault();
            navigate(e.detail.href);
          }}
          items={uniqueNames.map(name => ({
            type: 'link',
            text: name,
            href: name,
          }))}
        />
      }
    >
      <Box margin={{ top: 'm' }}>
        <Routes>
          {uniqueNames.map(name => (
            <Route
              path={`/${name}`}
              key={name}
              element={<SelectedDoc definitions={definitions} selected={name} onNavigate={onNavigate} />}
            />
          ))}
        </Routes>
      </Box>
    </PageLayout>
  );
}

function SelectedDoc({
  definitions,
  selected,
  onNavigate,
}: {
  definitions: { name: string; folder: string; promise: Promise<Definition> }[];
  selected: string;
  onNavigate: (path: string) => void;
}) {
  const [selectedDefinitions, selectDefinitions] = useState<Definition[]>([]);

  useEffect(() => {
    const matched = definitions
      .filter(({ name }) => name === selected)
      .sort((a, b) => a.folder.localeCompare(b.folder));

    if (matched.length > 0) {
      Promise.all(matched.map(it => it.promise)).then(selectDefinitions);
    }
  }, [definitions, selected]);

  return (
    <SpaceBetween size="m">
      {selectedDefinitions.map((definition, index) => (
        <LayoutNode key={index} definition={definition} onNavigate={onNavigate} level={1} />
      ))}
    </SpaceBetween>
  );
}

function LayoutNode({ definition, onNavigate, level }: NodeProps) {
  switch (definition.nodeType) {
    case 'Section':
      return <SectionNode definition={definition} onNavigate={onNavigate} level={level} />;
    case 'Table':
      return <TableNode definition={definition} onNavigate={onNavigate} level={level} />;
    case 'Title':
      return <TitleNode definition={definition} onNavigate={onNavigate} level={level} />;
    case 'Comment':
      return <CommentNode definition={definition} onNavigate={onNavigate} level={level} />;
    case 'Code':
      return <CodeNode definition={definition} onNavigate={onNavigate} level={level} />;
    case 'Literal':
      return <LiteralNode definition={definition} onNavigate={onNavigate} level={level} />;
    case 'Empty':
      return <EmptyNode />;
    default:
      return null;
  }
}

function SectionNode({ definition, onNavigate, level }: NodeProps) {
  const ref = useRef<HTMLDivElement>(null);
  const location = useLocation();

  const title = <LayoutNode definition={definition.title} onNavigate={onNavigate} level={level} />;
  const anchor = definition.title?.name ?? undefined;
  const heading =
    level === 1 ? (
      <Header variant="h2" id={anchor}>
        {title}
      </Header>
    ) : level === 2 ? (
      <Header variant="h3" id={anchor}>
        {title}
      </Header>
    ) : (
      <Box fontWeight="bold" id={anchor}>
        {title}
      </Box>
    );

  useEffect(() => {
    if ('#' + anchor === location.hash && ref.current) {
      ref.current.scrollIntoView();
    }
  }, [location.hash, anchor]);

  return (
    <div>
      {anchor && level < 3 ? <div ref={ref}>{heading}</div> : heading}

      <SpaceBetween size="m">
        {definition.items.map((item: Definition, index: number) => (
          <LayoutNode key={index} definition={item} onNavigate={onNavigate} level={level + 1} />
        ))}
      </SpaceBetween>
    </div>
  );
}

function TableNode({ definition, onNavigate, level = 1 }: NodeProps) {
  if (definition.items.length === 0) {
    return null;
  }

  const columnDefinitions = definition.header.map((name: string, index: number) => ({
    id: name,
    header: name,
    cell: (row: any) => <LayoutNode definition={row[index]} onNavigate={onNavigate} level={level + 1} />,
  }));

  return (
    <Table
      header={<Header>{definition.title}</Header>}
      items={definition.items}
      columnDefinitions={columnDefinitions}
      wrapLines={true}
      sortingDisabled={true}
    />
  );
}

function TitleNode({ definition }: NodeProps) {
  const kind = definition.kind !== 'Property' ? definition.kind : null;
  return (
    <SpaceBetween size="xs" direction="horizontal">
      {definition.flags.map((flag: string) => (
        <Badge key={flag}>{flag}</Badge>
      ))}
      {kind && <div>{kind}</div>}
      {definition.name}
    </SpaceBetween>
  );
}

function CommentNode({ definition }: NodeProps) {
  return (
    <SpaceBetween size="m">
      <Markdown>{definition.markdown}</Markdown>
      {definition.tags.map((tag: any) => (
        <CommentTag key={tag.name} {...tag} />
      ))}
    </SpaceBetween>
  );
}

function CommentTag({ name, markdown }: { name: string; markdown: string }) {
  return (
    <SpaceBetween size="s">
      <Box fontWeight="bold">{name}</Box>
      <Markdown>{markdown}</Markdown>
    </SpaceBetween>
  );
}

function CodeNode({ definition, onNavigate }: NodeProps) {
  return (
    <Code>
      {definition.tokens.map(([value, kind, options]: any, index: number) => (
        <CodeToken key={index} kind={kind} options={options} onNavigate={path => onNavigate(buildUrl(path))}>
          {value}
        </CodeToken>
      ))}
    </Code>
  );
}

function LiteralNode({ definition }: NodeProps) {
  return definition.value;
}

function EmptyNode() {
  return <Box>-</Box>;
}

function buildUrl(path: string[]) {
  return path.filter(entry => !entry.startsWith('@')).join('#');
}
