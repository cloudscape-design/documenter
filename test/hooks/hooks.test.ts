// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { describe, expect, test } from 'vitest';
import { buildCustomProject } from '../components/test-helpers';
import schema from '../../src/schema';
import { DeclarationReflection, ReflectionKind } from 'typedoc';

describe('Build code from type', () => {
  const project = buildCustomProject('../../fixtures/hooks/tsconfig.json', 'fixtures/hooks/*');

  ['useContainerQuery', 'useResizeObserver', 'useStableEventHandler'].forEach(name =>
    test(`Snapshot: ${name}`, () => {
      const reflection = project.findReflectionByName(name);

      const code: any = { returnType: '', parameters: [] };
      if (reflection && reflection.kind === ReflectionKind.Function) {
        const signature = (reflection as DeclarationReflection).signatures?.[0];
        code.returnType = schema.code.buildType(signature?.type);
        code.parameters = signature?.parameters?.map(p => ({ name: p.name, type: schema.code.buildType(p.type) }));
      }

      expect(code).toMatchSnapshot();
    })
  );
});
