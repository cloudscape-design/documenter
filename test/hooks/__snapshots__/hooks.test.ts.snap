// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Build code from type > Snapshot: useContainerQuery 1`] = `
{
  "parameters": [
    {
      "name": "mapFn",
      "type": "(entry: ContainerQueryEntry, prev: ObservedState | null) => ObservedState",
    },
    {
      "name": "deps",
      "type": "React.DependencyList",
    },
  ],
  "returnType": "[ObservedState | null,MutableRefObject<any>]",
}
`;

exports[`Build code from type > Snapshot: useResizeObserver 1`] = `
{
  "parameters": [
    {
      "name": "elementRef",
      "type": "ElementReference",
    },
    {
      "name": "onObserve",
      "type": "(entry: ContainerQueryEntry) => void",
    },
  ],
  "returnType": "void",
}
`;

exports[`Build code from type > Snapshot: useStableEventHandler 1`] = `
{
  "parameters": [],
  "returnType": "",
}
`;
